### ISSUE TITLE
kubectl port-forward flaky e2e tests

### ISSUE THREAD

**Issue by seans3** [kubernetes/kubectl#558]
**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):

**BUG REPORT**

**Test Grid Test**
https://gubernator.k8s.io/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-new-master-gci-kubectl-skew/11783#sig-cli-kubectl-port-forwarding-k8sio-with-a-server-listening-on-localhost-should-support-forwarding-over-websockets

**Kubernetes version** (use `kubectl version`):
Client: 1.11
Server: 1.11

**What happened**:

Test Grid test fails intermittently.

**FLAKE 1**

 It looks like a **RACE CONDITION**. The failure looks like:

```
Nov  6 14:40:52.585: INFO: >>> kubeConfig: /workspace/.kube/config
STEP: Creating the pod
STEP: Sending the expected data to the local port
STEP: Reading data from the local port
STEP: Verifying logs
Nov  6 14:41:09.010: INFO: Missing "^Accepted client connection$" from log: 
Nov  6 14:41:09.190: INFO: Pod log:
Accepted client connection
Received expected client data
```

This appears to be a race condition because it looks like its checking for `"^Accepted client connection$"`, then that missing statement happens 180 ms later.

**FLAKE 2**

There are multiple flakes with the following error:

`Failed to parse kubectl port-forward output: Forwarding from [::1]:34011 -> 80`

Log context is:

```
STEP: Creating the target pod
STEP: Running 'kubectl port-forward'
Nov  5 11:13:59.004: INFO: starting port-forward command and streaming output
Nov  5 11:13:59.004: INFO: Asynchronously running '../../../../kubernetes_skew/cluster/kubectl.sh ../../../../kubernetes_skew/cluster/kubectl.sh --server=https://35.202.150.113 --kubeconfig=/workspace/.kube/config port-forward --namespace=e2e-tests-port-forwarding-2xcwq pfpod :80'
Nov  5 11:13:59.005: INFO: reading from `kubectl port-forward` command's stdout
Nov  5 11:13:59.244: INFO: Failed to parse kubectl port-forward output: Forwarding from [::1]:34011 -> 80

Nov  5 11:13:59.260: INFO: Pod log:

[AfterEach] [sig-cli] Kubectl Port forwarding
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:142
STEP: Collecting events from namespace "e2e-tests-port-forwarding-2xcwq
```

A link to the failure is:  https://gubernator.k8s.io/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-new-master-gci-kubectl-skew/11756#sig-cli-kubectl-port-forwarding-k8sio-with-a-server-listening-on-0000-k8sio-that-expects-a-client-request-should-support-a-client-that-connects-sends-no-data-and-disconnects

The following PR seems very relevant to this flake: https://github.com/kubernetes/kubernetes/pull/46517

**What you expected to happen**:

Test Grid test to pass deterministically

**What else**

`kubectl port-forward` e2e test is at: https://github.com/kubernetes/kubernetes/blob/master/test/e2e/kubectl/portforward.go

/kind flake
/sig cli
/area kubectl
/priority P2
/assign goblain
/assign m1kola

**Comment by m1kola** [comment]
I didn't manage to run e2e tests locally (having troubles spinning up cluster of a specific version with `kubetest` because my upload speed isn't great), but I think I found the reason of flake 2. Flake 1 seems to be unrelated.

Flake 2-like fails seem to happen only on skew tests when we run latest `kubectl` against older version of the server (`skew-cluster-stable1-kubectl-latest-gke`, `skew-cluster-stable1-kubectl-latest-gce`).

I believe this happens because of this piece of code:

```go
	addresses = make([]listenAddress, len(parsed))
	id := 0
	for _, v := range parsed {
		addresses[id] = v
		id++
	}
```

where we iterate over `parsed` which is a map. Order in this case is not guaranteed. The older expect output to be like this:

```
Forwarding from 127.0.0.1:8000 -> 80
Forwarding from [::1]:8000 -> 80
```
Exactly in this order. But sometimes output (we can't see it form CI logs) appears like this:

```
Forwarding from [::1]:8000 -> 80
Forwarding from 127.0.0.1:8000 -> 80
```


New tests don't fail because they were updated from this:

```go
portForwardRegexp = regexp.MustCompile("Forwarding from 127.0.0.1:([0-9]+) -> 80")
```

to this:

```go
portForwardRegexp = regexp.MustCompile("Forwarding from (127.0.0.1|\\[::1\\]):([0-9]+) -> 80")
```

I'll try to fix it today-tomorrow.

**Comment by m1kola** [comment]
I prepared a PR (#70931), but then realised that failures have nothing to do with order of these messages... I think, it's still a good idea to merge my PR.

**Comment by m1kola** [comment]
Fix for flake 1 - https://github.com/kubernetes/kubernetes/pull/71048

**Comment by m1kola** [comment]
I have an update about flake 2. I was right when I thought I was wrong 🙂The reason tests are failing is because of ordering of output (explained in https://github.com/kubernetes/kubectl/issues/558#issuecomment-437682453).

I thought that my assumption is wrong, because we match port using [this regex](https://github.com/kubernetes/kubernetes/blob/v1.12.3/test/e2e/kubectl/portforward.go#L51-L53) which "doesn't care" about newlines or ordering, BUT the port forward command writes these two lines separately (two separate `fmt.Fprintf` calls):

```
Forwarding from [::1]:8000 -> 80
Forwarding from 127.0.0.1:8000 -> 80
```

And before trying to match port we do a **single** `Read` from port-forward's stdout [here](https://github.com/kubernetes/kubernetes/blob/v1.12.3/test/e2e/kubectl/portforward.go#L176-L178). Reader doesn't wait for the second line and returns what is available. So it will either be `Forwarding from 127.0.0.1:8000 -> 80` or `Forwarding from [::1]:8000 -> 80` depending on the order.

The above only happens in the skew tests because the e2e tests were updated in the original PR https://github.com/kubernetes/kubernetes/pull/46517, but not in the 1.12 branch.

---

We have two options to fix it:

1. Update 1.12 branch (we currently run skew tests against it)
2. Wait for skew tests to be switched to 1.13 which already has [updated regexp](https://github.com/kubernetes/kubernetes/blob/25b853393d1431a9686aebeea0568c261f0ef6b2/test/e2e/kubectl/portforward.go#L56-L58). Not sure how soon it is going to happen.

I'll look for an advice with sig-cli and report back on this as soon as possible.

**Comment by m1kola** [comment]
The following PRs were merged:

* https://github.com/kubernetes/kubernetes/pull/71048 - fixes flake 1 in the 1.13 branch
* https://github.com/kubernetes/kubernetes/pull/71621 - same as above, but for the 1.12 branch
* https://github.com/kubernetes/kubernetes/pull/71631 - fixes flake 2 in the 1.12 branch (1.13 already has a fix from the original PR).

Testgrid looks much better now. There are other PRs I opened while working on this issue, but they are not relevant to flakes (anymore).

I think, we can close this issue.

/close
